File: Programmers_Python/lv2_광물 캐기.py
import heapq


def solution(picks, minerals):
    # 곡갱이가 부족할 경우 캘수 있는 광물 만큼 조정
    if sum(picks) * 5 < len(minerals):
        minerals = minerals[:sum(picks) * 5]

    heap = []
    mapping = {"diamond": 25, "iron": 5, "stone": 1}
    value_minerals = [mapping[item] for item in minerals]

    for i in range(0, len(minerals), 5):
        group_sum = sum(value_minerals[i:i + 5])
        heapq.heappush(heap, (-group_sum, i))

    kind = 0
    fatigue = 0
    while sum(value_minerals) != 0:
        cur = heapq.heappop(heap)
        while picks[kind] == 0:     # 주의! 연속으로 도구가 없을 수 있으므로 도구 만날때까지
            kind += 1
        power = 5 ** (2 - kind)
        for i in range(cur[1], min(cur[1] + 5, len(minerals))):
            quo = value_minerals[i] // power
            fatigue += 1 if quo == 0 else quo
            value_minerals[i] = 0
        picks[kind] -= 1
    return fatigue


picks = [0, 0, 1]
minerals = ["diamond", "diamond", "diamond", "diamond", "diamond", "iron", "iron", "iron", "iron", "iron", "diamond"]
print(solution(picks, minerals))

# 광물 별 가중치를 매긴다음 광물을 5개씩 묶어서 나타냄
# step1) 캘 수있는 광물 길이 조정
# step2) 광물리스트 점수로 치환 후, 5묶음씩의 함 최대힙 넣기
# step3) 최대값을 우선 뽑은 후, 다이아부터 해결

# 5묶음씩 묶는거에서 너무 시간낭비함,,, 깔끔하게 for문 이용하기

File: Programmers_Java/Programmers_Java/src/lv2_광물캐기.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;

public class lv2_광물캐기 {
    public static void main(String[] args) {
        int[] picks = {0, 1, 1};
        String[] minerals = {"diamond", "diamond", "diamond", "diamond", "diamond", "iron", "iron", "iron", "iron", "iron", "diamond"};
        System.out.println(solution(picks, minerals));
    }
    public static int solution(int[] picks, String[] minerals) {
        // 광물 길이 조정하기
        if (Arrays.stream(picks).sum()*5 < minerals.length) {
            minerals = Arrays.copyOfRange(minerals, 0, Arrays.stream(picks).sum()*5);
        }

        // 광물에 따라 값 매기기
        int[] minerals_val = new int[minerals.length];
        HashMap<String, Integer> mapping = new HashMap<>(){{
            put("diamond", 25);
            put("iron", 5);
            put("stone", 1);
        }};

        for (int i = 0; i < minerals.length; i++) {
            minerals_val[i] = mapping.get(minerals[i]);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> b[1]-a[1]);
        for (int i = 0; i < minerals.length; i += 5) {
            int sum = 0;
            for (int j = i; j < Math.min(i + 5, minerals_val.length); j++) {
                sum += minerals_val[j];
            }
            pq.add(new int[]{i, sum});
        }

        int pick = 0;
        int fatigue = 0;
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            while (picks[pick] == 0) {
                pick++;
            }
            for (int i = cur[0]; i < Math.min(cur[0] + 5, minerals_val.length); i++) {
                int quo = (int) (minerals_val[i] / (Math.pow(5,2-pick)));
                fatigue += quo==0 ? 1 : quo;
            }
            picks[pick]--;
        }
        return fatigue;
    }
}
